Decoupled Notification From Widget

prateek-singh-3212 opened the pull request:
## Pull Request template

## Purpose / Description
 Notification is tightly coupled with SmallWidget status which results in  Notification whenever widget updates.

## Fixes
Fixes #8114 
Fixes #6476

## Approach
We can make the new repository which holds the meta Data (or Meta information) of all decks like Total deck cards, Total New cards to Review, Total Learning cards etc in sqlite Database and whenever the data is required by widget , Notification or any other activity, Service.

This discussion is documented in this PR ([this](https://github.com/ankidroid/Anki-Android/pull/5642)). 

### Flow Chart of Above Explained.

![untitled](https://user-images.githubusercontent.com/76490368/128639673-1a948cea-980d-4bfa-a6d2-7da5c5670d9f.png)

Flow Chart of Above Explained.

## How Has This Been Tested?

Firefox browser and Galaxy M51 (API 30) Aspect ratio 20:9

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -62,4 +62,47 @@ public boolean isDyn() {
+     * Time when deck is last modified.
+     * @return time in epoch
+     * */
+    public Long lastModified() {
```
Can you add these as functions outside `libAnki` and add a couple of tests. I'm in the process of migrating this class to Rust: https://github.com/ankidroid/Anki-Android/pull/9392

prateek-singh-3212 replied:
Sure, I will fix this soon.

prateek-singh-3212 replied:
I am not including these function in this PR. I will add this in future when I will work on Deck Reminders.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/DecksMetaData.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class DecksMetaData {
```
Please move out of `libAnki`, this isn't related to Anki Desktop

david-allison-1 replied:
It'd be great to see this in Kotlin, rather than Java

prateek-singh-3212 replied:
Okay, I will change it to kotlin

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/DecksMetaData.java:
```diff
@@ -0,0 +1,63 @@
+            total.addRev(node.getRevCount());
+        }
+        int eta = mCollection.getSched().eta(total, false);
+        Log.d("AABB", "getAllDeckNames: " + total.count() + " " + eta);
```
```suggestion
        Timber.d("getAllDeckNames: %d %d", total.count(), eta);
```

prateek-singh-3212 replied:
Oh! Forgot to fix that. Sorry 😅

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.java:
```diff
@@ -73,68 +60,26 @@ public static void update(Context context) {
-            if (sSmallWidgetEnabled) {
-                new AnkiDroidWidgetSmall.UpdateService().doUpdate(context);
-            }
-            Intent intent = new Intent(NotificationService.INTENT_ACTION);
```
Where has this code moved to?

prateek-singh-3212 replied:
I removed this part of code because it is the main cause of showing notification whenever widget updates.
And I think that this should not be the right place to put this part of code (saying for notification service).

I am planning to use workmanager and would love to hear you opinion on the same.

prateek-singh-3212 replied:
> this seems to remove functionality (widgets cause notifications) with no replacement

Can you elobarte? I dont know about this functionality.

I removed this part of code because it is the main cause of showing notification whenever widget updates.
And I think that this should not be the right place to put this part of code (saying for notification service).

I am planning to use workmanager and would love to hear you opinion on the same

david-allison-1 replied:
> I removed this part of code because it is the main cause of showing notification whenever widget updates.
And I think that this should not be the right place to put this part of code (saying for notification service).

We shouldn't be removing functionality without adding in a replacement, if we release the app after the PR is merged, then it's a regression in functionality

prateek-singh-3212 replied:
> WorkManager - I'm not familiar enough to have a good opinion.

Okay, No issues I am asking this because we are using Async Task which is deprecated.

I have added this part of code back because this part of code also handles the notification when widget is not shown.

prateek-singh-3212 replied:
@david-allison-1 moved this to AnkiWidigtSmall.class

david-allison-1 reviewed (changes requested):
Thanks!

This could do with more documentation, either by splitting out the commits, or a message in the commit message and PR summary.

I'll need to reread this after the comments are added, but from scanning, this seems to remove functionality (widgets cause notifications) with no replacement

prateek-singh-3212 commented:
**Issues**
1. Widget is not updating when app is not opened or in not present in background (real-time).
2. Widget takes 4-5 sec to update whenever their is update in data of due cards and if we kill the app while widget is updating (before 4-5 sec) then it will not show the updated data.

david-allison-1 commented:
To move this along: we either need to re-instate the widget based logic, or provide a better replacement

prateek-singh-3212 commented:
Yes, I also think so that we have to re-engineer the widget part . Should I create a new issue ?

david-allison-1 commented:
> Yes, I also think so that we have to re-engineer the widget part . Should I create a new issue ?

I thought we'd discussed in the issue that the widget based logic would need to be improved/replaced?

prateek-singh-3212 commented:
> I thought we'd discussed in the issue that the widget based logic would need to be improved/replaced?

I am unable to find any discussion on widget based logic. There is only one issue related to this i.e #6476. 
And this PR solves this (#6476) issue also by removing this part ⬇️ 
https://github.com/ankidroid/Anki-Android/blob/8ada06b2e6521aa4e9eff2a0539dc192ef89a894/AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.java#L114

david-allison-1 commented:
The widget-based code is known to be buggy, and better alternatives exist.

But, people are using the widget to get some notification-based functionality in AnkiDroid, and this removes this possibility without providing an alterative. 

I don't think we should remove this functionality for users until we have a better alternative.

prateek-singh-3212 commented:
@david-allison-1 I will refactor the widget part first.

prateek-singh-3212 commented:
**REFACTORED WIDGET**
✅ Notification will come irrespective of widget.
✅ Implemented Alarm Manager to update the widget which will improve the battery drain.
✅ Improved the feature which shows the notification whenever widget updates.

**WIDGET WORKING** 
Widget will work completely independently from the app. Widget will auto update itself after a fixed interval of time (for now 30 min). If widget is able to fetch data then it show it else no data will be shown (i.e empty widget but rare case when `DeckDueTreeNode ` is unable the send data.

**WIDGET NOTIFICATION**
There is feature which shows the notification whenever widget updates.  So I refactored it by storing the card no. when widget is last updated. If the new due cards is not equal to last updated due cards then it will show the notification else no change will be their. If we want to update the widget whenever app closes then we have to implement a callback which starts the Update notification service (I will try this  in  next commit).

**DONE TODO**
> TODO @mikehardy  - we can reduce battery usage by widget users by removing updatePeriodMillis from metadata
>       and replacing it with an alarm we set so device doesn't wake to update the widget, see:
>       https://developer.android.com/guide/topics/appwidgets/#MetaData

**LEARNING RESOURSE**
1. https://proandroiddev.com/android-alarmmanager-as-deep-as-possible-909bd5b64792
2. https://developer.android.com/guide/topics/appwidgets

prateek-singh-3212 commented:
Commit : *9f38230**
```
CardBrowser.java
DeckPicker.java
ModelBrowser.java
ModelFieldEditor.java
NoteEditor.java
Reviewer.java
StudyOptionsActivity.java
```

Removed Widget.update() from on Stop of these activities.

**TESTING PROCESS**
1. For now I have set alarm repeat time to `15 sec`. for both Notification Alarm and Notification Widget.
2. For ease of logging I have given the tag `AACC` for all the changes. I have implemented.
3. Test first without widget and after adding widget.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DecksMetaData.kt:
```diff
@@ -0,0 +1,54 @@
+     * **first** is due cards. <br></br>
+     * **second** is Expected time.
+     */
+    fun getTotalDueCards(): Pair<Int, Int> {
```
Let's use a data class here.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DecksMetaData.kt:
```diff
@@ -0,0 +1,54 @@
+
+class DecksMetaData(context: Context) {
+    private val mDecks: Decks
+    private val mCollection: Collection = CollectionHelper.getInstance().getCol(context)
```
Using `CollectionHelper` means that there's a chance this class will crash on init if the collection is closed or can't be loaded. I don't think this is what we want, and the collection (or a lambda to create the collection) would be better as a dependency

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/MetaDB.java:
```diff
@@ -438,12 +438,12 @@ public static void storeLookupDictionary(Context context, long did, int dictiona
-     * @return [due, eta]
-     */
+     * @return [due, eta] <br>
+     *<b> Widget Status table will be removed from database in future.</b>
```
Consider a `@WidgetCleanup(String)` annotation, we can then remove the annotation when the cleanup has been performed, and we can use "Find Usages" to get all instances where we need to perform cleanup

See: `RustCleanup` for a current implementation of this

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java:
```diff
@@ -41,49 +44,82 @@
     @Override
     public void onReceive(Context context, Intent intent) {
         Timber.i("NotificationService: OnStartCommand");
+        Log.d("AACC", "onReceive: NOTIFICATION");
```
```suggestion
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/services/NotificationService.java:
```diff
@@ -41,49 +44,82 @@
+        switch (intent.getIntExtra(CARD_NOTIFICATION_TYPE, CARD_NOTIFICATION_ALARM)) {
+            case CARD_NOTIFICATION_WIDGET: {
+                // We have to show the notification always. Because we are checking the condition before sending.
+                Log.d("AACC", "onReceive: NOTIFICATION WIDGET");
```
Use Timber, not `log`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/Deck.java:
```diff
@@ -62,4 +62,4 @@ public boolean isDyn() {
         return getInt("dyn") == Consts.DECK_STD;
     }
-}
+}
```
change can be reverted

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/widget/AnkiDroidWidgetSmall.java:
```diff
@@ -59,6 +62,7 @@ public void onEnabled(Context context) {
         Timber.d("SmallWidget: Widget enabled");
         SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(context);
         preferences.edit().putBoolean("widgetSmallEnabled", true).commit();
+        new  WidgetAlarm().setAlarm(context.getApplicationContext());
```
```suggestion
        new WidgetAlarm().setAlarm(context.getApplicationContext());
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/widget/WidgetAlarm.kt:
```diff
@@ -0,0 +1,63 @@
+        AnkiDroidWidgetSmall.UpdateService().doUpdate(context)
+    }
+
+    @SuppressLint("UnspecifiedImmutableFlag")
```
Let's fix this

david-allison-1 reviewed (changes requested):
This is looking better!

Could you split this out into more commits.

I'll come back to this tomorrow with a fresh pair of eyes (there's more to review), I'm finding it difficult to understand the context of the refactorings, vs the functional change, and more atomic commits would help here

prateek-singh-3212 commented:
![Untitled Diagram](https://user-images.githubusercontent.com/76490368/131218316-2be82573-71c4-437d-ba57-95997c5226d2.png)
@mikehardy @david-allison-1 Please review the plan.

david-allison-1 commented:
I don't think we want to reuse the timer. I feel we'd be better with two timers, as that reduces complexity further as it removes the double query of the data store.

The most important thing to me is finding the mechanism to display the notification once the app's been stopped and likely isn't in the background (like how a messenger still shows messages if it's not running).

prateek-singh-3212 commented:
Okay, I Understood your point. You want to work me like this.

![Untitled Diagram(1)](https://user-images.githubusercontent.com/76490368/131231874-a1205333-8dfe-4bec-b9d4-50ec03dac3f1.png)


Major Drawback with this approach is that widget and notification will not be synchronous with each other.
eg. due card is 150 and notification alarm trigger which shows the notification of due card 150. 
After 30 sec Due cards count updates to 155. and widget alarm triggers and updates the widget. Then the widget will show 155 due cards where as notification is still showing 150 which confuses the user and user is not able to figure out correct due cards count. Thats why I Suggested the [above approach](https://github.com/ankidroid/Anki-Android/pull/9411#issuecomment-907621847).  

> The most important thing to me is finding the mechanism to display the notification once the app's been stopped and likely isn't in the background (like how a messenger still shows messages if it's not running).

This and above approach both will run in background when alarm triggers and is completely separate from complete app. 

> as that reduces complexity further as it removes the double query of the data store and the dispatch.

I can reduce the double fetching of data in [first approch](https://github.com/ankidroid/Anki-Android/pull/9411#issuecomment-907621847) with the help of storing the number of due cards shown in last notification/widget update and comparing it with current due card number. If both are same then don't trigger the services (in context of [first approcach](https://github.com/ankidroid/Anki-Android/pull/9411#issuecomment-907621847)

david-allison-1 commented:
I'm personally fine with that caveat (as long as it's documented, and until we get a user complaint). 

Once we do, I feel we want the dependency to be from NotificationService -> Widget, rather than a general dispatch method which goes to both.

I predict that most users using the widget only use it to enable notifications.

prateek-singh-3212 commented:
> I'm personally fine with that caveat

Okay. I will work according to new plan. 

> Once we do, I feel we want the dependency to be from NotificationService -> Widget, rather than a general dispatch method which goes to both.

Sure, we can plan this in future. (If we get complains from users?)

DELETING LAST 2 COMMITS

prateek-singh-3212 commented:
David Please review it once again. Done All things as per plan

mikehardy commented:
>  then checking each minute seems sensible

I'm not sure if this is an issue or not but If we recompute the counts as a poll, instead of doing this as an event-based thing we'll be put in the "you use too much power" penalty bucket and background services will be adaptively killed (or start-denied) by power algorithms on device and notifications won't work for anyone

This stuff is hard to get right, but the fundamental thing is that notifications and widget counts need to be updated (not delete+re-create - same notification id re-used and just posted with new content) via an event subscription of some sort for the cases where the user is actually altering counts, then they should have zero activity except a PendingIntent for the notification to fire at the scheduled time (if any).

david-allison-1 commented:
Is there documentation for the criteria (from the most strict OS) for being put in the 'low power' bucket? Is this "too many wakeups", or "using too much time when woken up"

----

From a scheduling perspective, we can calculate the next time deterministically:

Reviews/day learn: We only need to check once per day at the day cutoff threshold - no issue. BUT: We may want this on a separate notification channel which won't cause vibrations, nobody wants a 4AM wakeup call to do their cards

Intraday, except when the app is being used, all we care about is:

* "sub-day learn" cards, which depend on the current time. We can calculate the time that the notification would next need to be shown trivially.
* API usage

----

If we show "you have over 250 cards due", then we don't need to calculate the exact number, as we don't display it.

If the issue is "using too much time when woken up", when the app is running, we can calculate the next time to wake up in a background thread (event based, rather than database-based). This would add a lot of complexity to the app, so I'd rather avoid it

mikehardy commented:
The ranking of terrible Android implementations changes constantly, this site is useful: https://dontkillmyapp.com/

prateek-singh-3212 commented:
Sorry, I was a little busy from past few weeks.
@david-allison-1 can we get back to this PR? 

**Regarding the algorithm which calculates the next time of notification.** 
Please give me the computing points to make algorithm like if reviews/day < 3 then show next notification in 30 minutes which says "Hey! where are you?  Your 250 cards are pending for today" like that. 
My plan is this 
![Untitled Diagram drawio](https://user-images.githubusercontent.com/76490368/133957983-4d36bdec-5f37-47aa-be2f-9be0cb3623f6.png)

david-allison-1 commented:
Glad to see you back!

This seems rather complicated.

Why do we need three services here?

As for the criteria: I'm open to suggestions but for a sensible first step, I'd use:

* One at the start of the day, shown once
* Check every interval for new "learn" cards, shown once each time the learn card count is over the threshold AND a specified amount of time has passed since the last notification

----

* (out of scope for the first implementation as it needs discussions) - one at the end of the day, if the user hasn't done X (for me, if any reviews/news are still pending)

@mikehardy @Arthur-Milchior - pinging as this is off the top of my head and I haven't worked through consequences

david-allison-1 commented:
@prateek-singh-3212 Added the "needs a new dev" label as it went stable, let me know if you'll have the time to finish this one off

prateek-singh-3212 commented:
@david-allison-1 Sorry I am unable to proceed this further as I am busy in my college classes.

david-allison-1 commented:
No worries! Life takes priority.

krmanik commented:
@prateek-singh-3212 I think you are implementing in another PR https://github.com/ankidroid/Anki-Android/pull/11487. If you wish to continue this PR then its okay, otherwise current PR can be closed.

prateek-singh-3212 commented:
No this PR for reference only. I am impleting this in new PR.
